### PR TITLE
Add `Filter` type

### PR DESCRIPTION
**Pull request by benzaria**
Filters elements from an `Array` based on whether they match the given `Type`.
If `Type` is `Boolean`, it filters out `falsy` values like `Boolean(T)` does.

Strict controls whether strict or loose type comparison is used (defaults to loose).

### CODE REVIEW

File: source/filter.d.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+			: [
+				...HeadAcc,
+				...IfFilter<FilterType<Array_[number], Type, Strict>, Array_>,
+				...TailAcc,
+			]
+		: Array_ extends readonly [(infer First)?, ...infer Rest]
+			? _Filter<Rest, Type, Strict, [
+				...HeadAcc,
+				...IfFilter<FilterType<First, Type, Strict>,
+					'0' extends OptionalKeysOf<Array_> // TODO: replace with `IsOptionalKeyOf`.
```
Comment (sindresorhus) [inline source/filter.d.ts]: This can be done now.

File: source/filter.d.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+*/
+type IfFilter<T extends boolean, V extends UnknownArray> = [T] extends [true] ? V : [];
+
+/**
+Filters elements from an array based on whether they extend the given type.
+
+If `Type` is `Boolean`, it filters out `falsy` values like {@link Boolean `Boolean(T)`} does.
+
+Strict controls whether strict or loose type comparison is used (defaults to loose).
+
```
Comment (sindresorhus) [inline source/filter.d.ts:69]: Need proper usage examples.

**Comment by sindresorhus** [comment]
Needs to be added to readme and index.d.ts

**Comment by sindresorhus** [comment]
Make sure to go through the pull request yourself and the diff and find mistakes. Etc. AI can be helpful for this.

File: source/filter.d.ts (hunk 1 above)
Comment (benzaria) [reply source/filter.d.ts]: sure thing!

**Comment by benzaria** [comment]
should the name of `ObjectFilter` type be `FilterObject` instead.

File: source/filter.d.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	type T1 = Filter<[1, 2, 3 | 4, 3?, 4?], 3>;
+	//=> [3 | 4, 3?]
+
+	type T2 = Filter<[1, 2, 3 | 4, 3?, 4?], 3, {strict: true}>;
+	//=> [3?]
+
+	type T3 = Filter<{a: 1; b: 2; c: 3 | 4; d: 3?; e: 4?}; 3>;
+	//=> {c: 3 | 4; d: 3?}
+
+	type T4 = Filter<{a: 1; b: 2; c: 3 | 4; d: 3?; e: 4?}; 3, {strict: true}>;
```
Comment (benzaria) [inline source/filter.d.ts]: I will fix this example

**Comment by sindresorhus** [comment]
> should the name of ObjectFilter type be FilterObject instead.

`FilterObject`

File: source/filter.d.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	type T1 = Filter<[1, 2, 3 | 4, 3?, 4?], 3>;
+	//=> [3 | 4, 3?]
+
+	type T2 = Filter<[1, 2, 3 | 4, 3?, 4?], 3, {strict: true}>;
+	//=> [3?]
+
+	type T3 = Filter<{a: 1; b: 2; c: 3 | 4; d: 3?; e: 4?}, 3>;
+	//=> {c: 3 | 4; d: 3?}
+
+	type T4 = Filter<{a: 1; b: 2; c: 3 | 4; d: 3?; e: 4?}, 3, {strict: true}>;
```
Comment (sindresorhus) [inline source/filter.d.ts]:
```suggestion
	type T4 = Filter<{a: 1; b: 2; c: 3 | 4; d: 3?; e?: 4}, 3, {strict: true}>;
```

etc

File: test-d/object-filter.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+// Edge cases
+expectType<ObjectFilter<any, never>>({});
+expectType<ObjectFilter<never, any>>({} as never);
+expectType<ObjectFilter<never, never>>({} as never);
+
+expectType<ObjectFilter<{}, number>>({});
+expectType<ObjectFilter<{a: never; b: never}, number>>({});
+expectType<ObjectFilter<{a: never; b: never}, never>>({});
+expectType<ObjectFilter<{a: never; b: never}, never, {strict: true}>>({} as {a: never; b: never});
```
Comment (sindresorhus) [inline test-d/object-filter.ts]: some more tests:

```ts
// Readonly properties (verifies readonly modifiers are preserved)
expectType<ObjectFilter<{readonly a: 1; b: 2; readonly c: 3}, number>>({} as {readonly a: 1; b: 2; readonly c: 3});
expectType<ObjectFilter<{readonly a: string; b: number; readonly c: boolean}, string>>({} as {readonly a: string});
expectType<ObjectFilter<{readonly a?: string; b?: number; c: boolean}, string>>({} as {readonly a?: string});
```

Comment (benzaria) [reply test-d/object-filter.ts]: I totally missed to add cases for readonly props and arrays.

File: source/filter.d.ts (hunk 4 above)
Comment (benzaria) [reply source/filter.d.ts]: I fixed them.

**Comment by benzaria** [comment]
> > should the name of ObjectFilter type be FilterObject instead.
> 
> `FilterObject`

Just pointing out that we also could merge them in one `Filter` type.

**Comment by sindresorhus** [comment]
A single type is tempting. Any downsides?

**Comment by benzaria** [comment]
> A single type is tempting. Any downsides?

should not be any, I guess.

**Comment by benzaria** [comment]
@sindresorhus the merge of both types was flawless with no problems.

File: source/internal/array.d.ts [hunk 6]
```diff
@@ -87,6 +88,42 @@ Returns whether the given array `T` is readonly.
...
+type Pass2 = IsEmptyArray<readonly []>;
+//=> true
+
+type Fail1 = IsEmptyArray<[0]>;
+//=> false
+
+type Fail2 = IsEmptyArray<[0?]>;
+//=> false
+
+type Fail3 = IsEmptyArray<...string[]>;
```
Comment (sindresorhus) [inline source/internal/array.d.ts]: I don't think this is valid TS. Ensure all examples are correct.

File: source/internal/type.d.ts [hunk 7]
```diff
@@ -129,4 +130,37 @@ export type IsExactOptionalPropertyTypesEnabled = [(string | undefined)?] extend
...
+	false: ExtendsLoose<T, U>;
+}[`${S}`];
+
+/**
+Performs a loose type comparison: checks if any of the members in `T` extends `U`.
+
+This is useful when needing to know if any member of `T` extends `U` without returning `boolean`.
+*/
+export type ExtendsLoose<T, U> = IsNotFalse<T extends U ? true : false>;
+// ? Should this type get exposed publicly
```
Comment (sindresorhus) [inline source/internal/type.d.ts]: We can discuss that later if there is a need. No need for comment about it on every type.

File: source/internal/type.d.ts [hunk 8]
```diff
@@ -129,4 +130,37 @@ export type IsExactOptionalPropertyTypesEnabled = [(string | undefined)?] extend
...
+
+/**
+Union of `falsy` types in TypeScript. `never` isn't included.
+*/
+export type Falsy = 0 | 0n | '' | false | null | undefined; // `| never`
+
+/**
+Checks if `T` is **not** {@link Falsy `falsy`} similar to `Boolean(T)`.
+*/
+// TODO: expose this type publicly
```
Comment (sindresorhus) [inline source/internal/type.d.ts]:
```suggestion
```

File: source/internal/type.d.ts (hunk 7 above)
Comment (benzaria) [reply source/internal/type.d.ts]: For sure!

File: source/internal/array.d.ts (hunk 6 above)
Comment (benzaria) [reply source/internal/array.d.ts]: You're right, I fixed it.

**Comment by sindresorhus** [comment]
Some things:

1. The `Boolean` special-case is currently too broad.

Right now truthy filtering is being selected not only for `Boolean`, but also for wider filter types like `unknown` and `any`. That changes behavior in a surprisng way. Example: filtering `[0, "", false]` by `unknown` ends up as `[]`, when it should keep all values.

2. Union results ar being over-narrowed.

When filtering a union, branches that become empty are removed from the final result. That makes the output unsound for real union inputs.

Examples:
- `Filter<[1] | [2], 1>` resolves to `[1]`, but should be `[1] | []`.
- `Filter<{a: 1} | {b: 2}, 1>` resolves to `{a: 1}`, but should be `{a: 1} | {}`.

3. One strict object example in docs is incorrect/inconsistant with the rest of the examples.

4. Tests are missing coverage for the two cases above.

- filtering by `unknown`/`any` with falsy values,
- union branches where one side filters to an empty result.
